### PR TITLE
Refactor contracts, utils, helpers, sandbox

### PR DESCRIPTION
**Pull request by anxolin**
This is the first big PR for the refactor of dex-js

i've moved all the contract definitions from `dex-telegram` to `dex-js` the idea is to be able to share utils and the contracts between any module including `dex-react` and `dex-telegram`

I know some utils are better in `react-web`. In some of the following PRs we have to move those utils here so all modules can benefit from them.

Additionally, it adds `ttypescript` and `typescript-transform-paths` to be able to fix the paths in the transpiled files.

Also, it renames the contract to `BatchExchange` (although I still didn't update the smart contracts to it's last version). I plan to do this after we merge this PR and refactor `dex-react`.

This PR has a lot's of commits of publishing of release candidates to NPM (i.e. 0.1.0-RC2). This was to test it's usage in `dex-telegram`. So there will be a matching PR that removes all this logic from there:
- https://github.com/gnosis/dex-telegram/pull/63

### CODE REVIEW

File: src/contracts/BatchExchangeContract.ts [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+  buyToken: string
+  sellToken: string
+  validFrom: string
+  validUntil: string
+  priceNumerator: string
+  priceDenominator: string
+}
+
+export interface BatchExchangeContract extends Contract {
+  // constructor(jsonInterface: any[], address?: string, options?: ContractOptions)
```
Comment (W3stside) [inline src/contracts/BatchExchangeContract.ts]: remove?

File: src/contracts/BatchExchangeContract.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    getPendingDepositBatchNumber(user: string, token: string): TransactionObject<string>
+
+    withdraw(user: string, token: string): TransactionObject<void>
+
+    MAX_TOUCHED_ORDERS(): TransactionObject<string>
+
+    addToken(token: string): TransactionObject<void>
+
+    placeValidFromOrder(
+      buyToken: number | string,
```
Comment (W3stside) [inline src/contracts/BatchExchangeContract.ts:110]: could be better to create a type from `number | string` - your call, I'm fine either way

File: src/contracts/Erc20Contract.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+import { TransactionObject } from 'contracts/types'
+import { Contract } from 'web3-eth-contract'
+
+export interface Erc20Contract extends Contract {
+  clone(): Erc20Contract
+
+  methods: {
+    totalSupply(): TransactionObject<string>
+    decimals(): TransactionObject<string>
```
Comment (W3stside) [inline src/contracts/Erc20Contract.ts:11]: do `constants` return a `TransactionObject`? thought they were just `Promise<string>`s

File: src/contracts/index.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import assert from 'assert'
+import { web3 } from 'helpers/web3'
+import { BatchExchangeContract } from './BatchExchangeContract'
+import { Erc20Contract } from './Erc20Contract'
+
+function getStableConverterContract (): BatchExchangeContract {
+  const { STABLE_COIN_CONTRACT_ADDRESS } = process.env
+  assert(STABLE_COIN_CONTRACT_ADDRESS, 'STABLE_COIN_CONTRACT_ADDRESS env is required')
+
+  const stableCoinContractAddress = STABLE_COIN_CONTRACT_ADDRESS as string
```
Comment (W3stside) [inline src/contracts/index.ts:10]: why is this required? shouldn't the `assert` tell TS that it's indeed a `string`?

File: src/contracts/types.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+type CallTxCallback = <T>(error: Error | null, result: T) => void
+type SendTxCallback = (error: Error | null, transactionHash: string) => void
+type EstimateGasTxCallback = (error: Error | null, gasAmount: number) => void
+
+export type Callback<T> = (error: Error, result: T) => void
+
+export interface TransactionObject<T, U extends any[] = []> {
+  arguments: U
+
+  call(callback: CallTxCallback): Promise<T>
```
Comment (W3stside) [inline src/contracts/types.ts:16]: Ah, i see. Disregard my comment about the return values of `constants` above

File: src/helpers/shutdown.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
+import Logger from 'helpers/Logger'
+import { Command, AsyncCommand } from '../types'
+
+const log = new Logger('helpers:shutdown')
+type QuitSignal = 'SIGINT' | 'SIGTERM' | 'SIGQUIT'
+const POSIX_SIGNALS: QuitSignal[] = ['SIGINT', 'SIGTERM', 'SIGQUIT']
```
Comment (W3stside) [inline src/helpers/shutdown.ts:5]: maybe move to a top level file like `const/signals` unless it's only used here

File: src/helpers/shutdown.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  await Promise.all(
+    listeners.map(listener => {
+      return listener()
+    })
+  )
+}
+
+function _doShutDown (reason: string) {
+  function _doExit (returnCode: number | undefined) {
+    log.debug('The app is ready to shutdown! Good bye! :)')
```
Comment (W3stside) [inline src/helpers/shutdown.ts]: Nit: `Goodbye`

Comment (W3stside) [reply src/helpers/shutdown.ts]: (one word)

File: src/helpers/shutdown.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    process.exit(returnCode)
+  }
+
+  shutDown(reason)
+    .then(() => {
+      _doExit(0)
+    })
+    .catch(error => {
+      log.error({
+        msg: 'Error while shut down the app: ' + error.toString(),
```
Comment (W3stside) [inline src/helpers/shutdown.ts]: Nit: `Error while shutting down app: `

File: src/helpers/web3.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import assert from 'assert'
+import Web3 from 'web3'
+
+import Logger from 'helpers/Logger'
+const log = new Logger('helpers:web3')
+const nodeUrl = process.env.NODE_URL
+
+log.info('Connecting to ethereum using web3: %s', nodeUrl)
+assert(nodeUrl, 'NODE_URL env var is required')
+export const web3: Web3 = new Web3(nodeUrl as string)
```
Comment (W3stside) [inline src/helpers/web3.ts:10]: `assert` should be catching the right type... is it the TS `assert` from `dex-js` or `node`?

File: src/contracts/Erc20Contract.ts (hunk 3 above)
Comment (Velenir) [reply src/contracts/Erc20Contract.ts:11]:
```js
txObj = Contract.methods.constant() (or whatever)
txObj.call() -> Promise<constant type>
txObj.send() -> tx PromiEvent
```

File: src/utils/ethereum.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
+import BN from 'bn.js'
+
+export { fromWei, toWei, isBN, toBN } from 'web3-utils'
+
+export function toBnOrNull (value: string | number): BN | null {
+  if (value === undefined || value === null || value === '') {
```
Comment (W3stside) [inline src/utils/ethereum.ts]: use `assert` here

File: src/contracts/index.ts (hunk 4 above)
Comment (Velenir) [reply src/contracts/index.ts:10]: Perhaps cause it's the old node assert, not like the one you wrote that `asserts is string`

File: src/contracts/types.ts (hunk 5 above)
Comment (Velenir) [reply src/contracts/types.ts:16]: This thing will also change with the new typechain

File: src/contracts/BatchExchangeContract.ts (hunk 2 above)
Comment (Velenir) [reply src/contracts/BatchExchangeContract.ts:110]: As we generate or will generate these types with typechain, it's not up to us

File: test/sandbox/contract/getAbi.ts [hunk 11]
```diff
@@ -1,16 +1,17 @@
...
+// TODO: Change to BatchExchange
 import { abi as abiItems } from '@gnosis.pm/dex-contracts/build/contracts/StablecoinConverter.json'
 require('dotenv').config()
 
 /**
  *  SANDBOX: Get ABI from the contracts
- *  RUN:     yarn sandbox test/sandbox/contracts/getAbi.ts
+ *  RUN:     yarn sandbox test/sandbox/contract/getAbi.ts
  */
 // const log = new Logger('sandbox:contracts:getAbi')
```
Comment (W3stside) [inline test/sandbox/contract/getAbi.ts]: remove?

File: test/sandbox/contract/tokenAddressToIdMap.ts [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+require('dotenv').config()
+
+/**
+ *  SANDBOX: Get token id from token addresses
+ *  RUN:     yarn sandbox test/sandbox/contract/tokenAddressToIdMap.ts
+ */
+const log = new Logger('sandbox:contract/tokenAddressToIdMap')
+
+async function exec (): Promise<void> {
+  const networkId = 4
```
Comment (W3stside) [inline test/sandbox/contract/tokenAddressToIdMap.ts]: is this meant to be hardcoded?

File: test/sandbox/web3/getPastLogs.ts [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+  assert(contractAddress, 'STABLE_COIN_CONTRACT_ADDRESS is required env var')
+  const topics = [
+    // OrderPlacement
+    // '0x951b66125eb3ef23450a52bb65cf4786c423cc944abb7de8a749a7d05642060e',
+
+    // OrderCancelation
+    // '0x36adb26c9fa7b5219b043ac78b698e6e539a937ff88758ab51b04d331e7111ed'
+
+    // Deposit
+    '0xdcbc1c05240f31ff3ad067ef1ee35ce4997762752e3a095284754544f4c709d7'
```
Comment (Velenir) [inline test/sandbox/web3/getPastLogs.ts:24]: How did you get these topics? `web3.eth.abi.encodeEventSignature`? Should we do that instead, so we don't have these magic numbers here?

**Review by Velenir — APPROVED** [review]
Should we import `src/contracts/BatchExchangeAbi.json` and all other abi from `dex-contracts` maybe?

File: test/sandbox/contract/tokenAddressToIdMap.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+  log.info('Get token address for network id: %d', networkId)
+  const getAllTokenIds = tokenList.map(async token => {
```
Comment (W3stside) [inline test/sandbox/contract/tokenAddressToIdMap.ts]: from what I remember you can't await in a `map` @Velenir help me out here

So you should do:
```js
const getPromisedAllTokenIds = tokenList.map(token => {
    const { symbol, name, addressByNetwork } = token
    const tokenAddress = addressByNetwork[networkId]
    
     return batchExchangeContract.methods.tokenAddressToIdMap(tokenAddress).call().then(tokenId => log.info(`${name} (${symbol}): ${tokenAddress}  ---->  ${tokenId}`))
})

return Promise.all(getPromisedAllTokenIds)
```

**Review by W3stside — CHANGES_REQUESTED** [review]

File: test/sandbox/contract/tokenAddressToIdMap.ts (hunk 14 above)
Comment (Velenir) [reply test/sandbox/contract/tokenAddressToIdMap.ts]: `Array.map(async func) = Array<Promise>` You can await inside as long as you expect a promise in the end

**Review by W3stside — APPROVED** [review]

File: test/sandbox/contract/tokenAddressToIdMap.ts (hunk 14 above)
Comment (W3stside) [reply test/sandbox/contract/tokenAddressToIdMap.ts]: gotcha - efficiency wise are they the same? the `promise.all` vs `async map`?

Comment (W3stside) [reply test/sandbox/contract/tokenAddressToIdMap.ts]: (in this context)

Comment (Velenir) [reply test/sandbox/contract/tokenAddressToIdMap.ts]: There's no `vs` you still have to `Promise.all` them, like happens on L23

Comment (W3stside) [reply test/sandbox/contract/tokenAddressToIdMap.ts]: i mean promise all-ing outside the loop

Comment (W3stside) [reply test/sandbox/contract/tokenAddressToIdMap.ts]: yeah ok nvm! thanks

File: src/contracts/BatchExchangeContract.ts (hunk 1 above)
Comment (anxolin) [reply src/contracts/BatchExchangeContract.ts]: will do

File: src/contracts/index.ts (hunk 4 above)
Comment (anxolin) [reply src/contracts/index.ts:10]: good point, maybe this can be refactored so we move the assert here in further prs

File: src/contracts/types.ts (hunk 5 above)
Comment (anxolin) [reply src/contracts/types.ts:16]: Yep, for now is just refactoring, I didn't want to make the changes yet, so it would be more difficult to review

File: src/helpers/shutdown.ts (hunk 7 above)
Comment (anxolin) [reply src/helpers/shutdown.ts]: yes, right. thx

File: src/helpers/web3.ts (hunk 9 above)
Comment (anxolin) [reply src/helpers/web3.ts:10]: It's from node. I think it's good if we use yours. I'll open an issue to address it in further PRs.

Comment (anxolin) [reply src/helpers/web3.ts:10]: https://github.com/gnosis/dex-js/issues/29

File: test/sandbox/contract/tokenAddressToIdMap.ts (hunk 12 above)
Comment (anxolin) [reply test/sandbox/contract/tokenAddressToIdMap.ts]: yep, not an issue. It's just a sandbox

image

**Comment by anxolin** [comment]
> Should we import src/contracts/BatchExchangeAbi.json and all other abi from dex-contracts maybe?

dex-contracts will publish their own contracts typscrypt definition, so I was thinking on reviewing that when they get to do that.

File: test/sandbox/web3/getPastLogs.ts (hunk 13 above)
Comment (anxolin) [reply test/sandbox/web3/getPastLogs.ts:24]: This is a sandbox, I agree we could do that. But it's testing get past logs in isolation, so you can use any topic.

There's another sandbox to get the topics, you can get it from etherscan too. It's not meant to be "application code" this is why I would find it ok

image

File: src/contracts/index.ts (hunk 4 above)
Comment (anxolin) [reply src/contracts/index.ts:10]: https://github.com/gnosis/dex-js/issues/29

File: test/sandbox/web3/getPastLogs.ts (hunk 13 above)
Comment (Velenir) [reply test/sandbox/web3/getPastLogs.ts:24]: cool

File: src/contracts/BatchExchangeContract.ts [hunk 15]
```diff
@@ -0,0 +1,163 @@
+import { Contract, EventOptions } from 'web3-eth-contract'
+import { TransactionObject, ContractEvent, Callback } from './types'
+import { EventLog } from 'web3-core'
+// import { AbiItem } from 'web3-utils'
+// import { Contract, EventOptions, ContractOptions } from 'web3-eth-contract'
```
Comment (alfetopito) [inline src/contracts/BatchExchangeContract.ts]: Dead code?

**Review by alfetopito — APPROVED** [review]
